articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,247 @@
+
+**Proposal:** Change request for `rosdep` to consider information beyond the platform to determine the mapping.
+
+**Decision:** With that change the `rosdep` rules should share a single namespace (this is currently already the case).
+
+### Package manifest
+
+In ROS 1 two versions of the package manifest are supported: format 1 specified in [REP 127](http://www.ros.org/reps/rep-0127.html) and format 2 specified in [REP 140](http://www.ros.org/reps/rep-0140.html).
+While ROS 2 only supports format 2 there is no technical limitation behind that design decision.
+It simply reduces the effort to distinguish the different semantics of some depend tags in various parts of the software.
+
+It might be the case that a to-be-specified format 3 will be created to add support for the desired features to describe [dependency groups]() as well as declaring an [ABI version]().
+
+#### ROS dependencies
+
+While a ROS 1 package might depend on `roscpp` and `message_generation` a ROS 2 package depends on `rclcpp` and `rosidl_message_generation` instead.
+
+#### Build type
+
+Almost all build types for ROS 1 and ROS 2 are disjoint except the build type `cmake` which can be processed by both build tools.
+Since it is not desired to change existing `catkin` / `ament_cmake` packages to use the build type `cmake` those are currently different for each ROS version.
+
+### CMake version
+
+ROS 2 requires CMake version 3.5+.
+It should be easily possible to update existing CMake code to be compatible with various versions of CMake.
+
+### Python version
+
+ROS 2 requires Python version 3.5+.
+It should be easily possible to update existing Python code to be compatible with Python 2.7 as well as Python 3.5+.
+
+### CMake API
+
+ROS 1 packages use the CMake API provided by [catkin](https://github.com/ros/catkin).
+ROS 2 packages use a slightly different CMake API provided by [ament_cmake](https://github.com/ament/ament_cmake).
+The rational for the changes and the different name are described in [this article](http://design.ros2.org/articles/ament.html).
+
+**Idea:** Create a ROS 2 package names `catkin` which provides the same CMake API as the ROS 1 package (minus unavailable features like the devel space) by mapping the calls to `ament_cmake` API.
+In addition it is expected that existing packages need to call a new function at the end of their CMake code to account for the fact that `ament_package` needs to be invoked *after* all targets where `catkin_package` needs to be invoked *before* them.
+
+### Message / service definitions
+
+In ROS 2 message / service definitions are expected to be in the subdirectory `msg` / `srv`.
+This implicitly ensures that message and service names are unique.
+There is no technical reason behind that design decision rather than simplifying the implementation.
+
+**Idea:** Allow arbitrary locations in the source space.
```

Comment:
I see no mention of a benefit.

Reply:
While arbitrary locations are possible in ROS 1 in ROS 2 we simplified the implementation by just allow one locations. It has been requested before to make this more flexible again (ros2/rosidl#213).

Reply:
Keeping the restriction on message and service locations may make it easier for new users to learn ROS 2 because it means there can be a tutorial with a canonical ROS2 package layout.

Even though it's not the case, it feels like the same restrictions are in ROS 1. `genmsg` `add_message_files` expects messages to be in `msg` by default. It allows changing the directory, but I don't think that's easy to discover. For a long time I thought this was a restriction and thought the beginner tutorials should say the messages must be placed there.

Reply:
I think the ROS 1 API is reasonable for this: use a sane default which is recommended in tutorials but allow the user to provide a custom location if necessary.

Reply:
I'm in favour of having a canonical package layout in ROS 2 that is taught in tutorials and highly recommended to be followed. It makes developing tools easier.

I don't think it needs to be absolutely enforced; having the ability to provide a custom location as @dirk-thomas said is sensible. I think it should be implied that putting them there is the best approach and you put them elsewhere at your own (hopefully informed) risk of someone's tool not working.